### PR TITLE
reboot: Fix typo and support bare Linux systems

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
This fixes a problem for bare Linux systems that do not support 'who -b' or 'uptime -s'.
For example, OpenELEC does not ship with `who` and does not support `uptime -s`.

Probably needs a backport to Linux v2.7.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
reboot

##### ANSIBLE VERSION
v2.8 and v2.7

### CODE REVIEW

File: lib/ansible/plugins/action/reboot.py [hunk 1]
```diff
@@ -103,9 +103,14 @@ def get_system_boot_time(self):
         if '1970-01-01 00:00' in command_result['stdout']:
             command_result = self._low_level_execute_command('uptime -s', sudoable=self.DEFAULT_SUDOABLE)
 
+        # This is a last resort for bare Linux systems (e.g. OpenELEC) where 'who -b' or 'uptime -s' are not supported.
+        # Other options like parsing /proc/uptime or default uptime output are less reliable than this
+        if command_result['rc'] != 0:
+            command_result = self._low_level_execute_command('cat /proc/sys/kernel/random/boot_id', sudoable=self.DEFAULT_SUDOABLE)
```
Comment (bcoca) [inline lib/ansible/plugins/action/reboot.py]: wouldn't we need the previous boot id to compare to?

File: lib/ansible/plugins/action/reboot.py [hunk 2]
```diff
@@ -103,9 +103,14 @@ def get_system_boot_time(self):
...
+
         if command_result['rc'] != 0:
             raise AnsibleError("%s: failed to get host boot time info, rc: %d, stdout: %s, stderr: %s"
-                               % (self._task.action, command_result.rc, to_native(command_result['stdout']), to_native(command_result['stderr'])))
+                               % (self._task.action, command_result['rc'], to_native(command_result['stdout']), to_native(command_result['stderr'])))
```
Comment (bcoca) [inline lib/ansible/plugins/action/reboot.py]: I would aggregate both errors here, both 'uptime' and 'cat /proc..'

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/action/reboot.py (hunk 1 above)
Comment (samdoran) [reply lib/ansible/plugins/action/reboot.py]: @bcoca The comparing is done in `check_boot_time()`. This method just gets the boot time and returns it.

Comment (samdoran) [reply lib/ansible/plugins/action/reboot.py]: I'm wondering if we should just switch to using this command rather than `who -b`, then `uptime-s`, then `cat ...`. Since this is Linux specific, it doesn't work on FreeBSD/Darwin/Solaris, so we would have to implement a lookup dict for boot time commands as well.

Comment (bcoca) [reply lib/ansible/plugins/action/reboot.py]: actually .. uptime is a bad check then, since you get a race condition in the case of 'chained reboots'

not something that will happen often, but you can get to the point in which the previous uptime < current uptime even though you rebooted. You should check the uptime vs the time since you started the reboot.

Comment (samdoran) [reply lib/ansible/plugins/action/reboot.py]: The `uptime -s` (and `who -b`) command return a string that is the time the last system was booted, different than what `uptime` is normally used for, which is reporting _total_ system up time.

The plugin doesn't do anything more than look for the output of this string to change. If it has not changed, it assumes the system has not been rebooted.

Since it is not doing `timedelta` comparison on the uptime output, I do not believe we will hit the scenario you describe.

Comment (bcoca) [reply lib/ansible/plugins/action/reboot.py]: i sit corrected then, i should have looked closer at the switches

File: lib/ansible/plugins/action/reboot.py (hunk 2 above)
Comment (samdoran) [reply lib/ansible/plugins/action/reboot.py]: Agree with this. Don't want to swallow the output from the initial command.

Long term, it feels like we may need to change distribution to being an attribute, that way we can run `cat ...` for Linux hosts, and use `who -b` for others, removing the need for `uptime -s` and a hard coded tree of commands to attempt.

**Review by samdoran — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/action/reboot.py (hunk 2 above)
Comment (dagwieers) [reply lib/ansible/plugins/action/reboot.py]: I fixed it so we collect all the stdout and stderr information in case we need it for the AnsibleError.

The problem with having only a single guaranteed command per OS is
1. You would need to know the platform/OS you're working with
2. I don't think it's that clear-cut

So this fallback method, while not necessarily 100% clean, it does the job effectively.
And `who -b` and `uptime -s` are very sensible things to do before resorting to the random boot_id. (Which may not be present in all Linux kernels either ?)

**Comment by dagwieers** [comment]
Please re-review !

File: lib/ansible/plugins/action/reboot.py (hunk 2 above)
Comment (samdoran) [reply lib/ansible/plugins/action/reboot.py]: My next question was going to be if the random boot ID is widely available. I guess I'll have to do some research to see when it was added.

I created a WIP PR (#45656) exploring what it would look like to add platform and dist to the plugin. I'm not terribly happy with it, but it's a start (and it accounts for Alpine, which is particularly annoying since it lacks both `who` and `uptime` for no good reason (IMO)).

That PR also changes to using the random boot ID by default rather than `who -b`. That has the nice side effect of being able to remove the `uptime -s` command for systems that lack a RTC, but at the cost of all the added complexity of platfrom and dist checking.

File: lib/ansible/plugins/action/reboot.py [hunk 3]
```diff
@@ -94,18 +94,31 @@ def construct_command(self):
...
         # Fall back to using uptime -s for those systems.
         # https://github.com/systemd/systemd/issues/6057
         if '1970-01-01 00:00' in command_result['stdout']:
+            stdout += command_result['stdout']
+            stderr += command_result['stderr']
             command_result = self._low_level_execute_command('uptime -s', sudoable=self.DEFAULT_SUDOABLE)
 
+        # This is a last resort for bare Linux systems (e.g. OpenELEC) where 'who -b' or 'uptime -s' are not supported.
+        # Other options like parsing /proc/uptime or default uptime output are less reliable than this
         if command_result['rc'] != 0:
```
Comment (samdoran) [inline lib/ansible/plugins/action/reboot.py:112]: One issue with this is that `win_reboot` uses this same method. So if the windows `DEFAULT_BOOT_TIME_COMMAND` results in an rc != 0, it will then try this `cat` command, which will gloriously fail.

The previous conditional which runs `uptime -s` won't ever match on Windows because that string will never match the output of `(Get-WmiObject -ClassName Win32_OperatingSystem).LastBootUpTime`.

This is the point where it would be helpful to have platform/dist as a class attribute, either as an additional check here, or to look up the correct boot time command based on platform/dist. I'm also open to other ideas.

**Review by samdoran — CHANGES_REQUESTED** [review]
Please add a changelog fragment as well.

File: lib/ansible/plugins/action/reboot.py (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/plugins/action/reboot.py:112]: I don't mind if a fallback method fails though. As long as we cover all bases so in the end the function always works. For win_reboot it could be a lot easier if we have a list of commands, for both Windows and Unix.

Comment (samdoran) [reply lib/ansible/plugins/action/reboot.py:112]: In my WIP PR, I did change the "get the last boot time" command to be a variable, so that will remove the hard coding here.

Will fix in future™.

Still, I can see it being confusing to someone running `win_reboot` (assuming the first command failed to get the last boot time) if they get this error:

```
win_reboot: failed to get host boot time info, rc: 1, stdout: 20180913204044.576906+000
, stderr: cat : Cannot find path 'C:\proc\sys\kernel\random\boot_id' because it does not exist.
At line:1 char:1
+ cat /proc/sys/kernel/random/boot_id
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (C:\proc\sys\kernel\random\boot_id:String) [Get-Content], ItemNotFoundEx
ception
+ FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetContentCommand
```

Evidently Windows has a `cat` command now!?

Comment (dagwieers) [reply lib/ansible/plugins/action/reboot.py:112]: Well, the distinction between Windows and Unix is now easy because it's a different entry-point. So we don't need to do a `cat` on Windows. That was my main point. But I also don't mind having the fallback mechanism on Unix, that first does `who -b`, then `uptime -s`, and then `cat /.../boot_id`. Which to me is quite sensible as well. It's a trade-off for simplicity and in most cases it only requires a single command `who -b`. If you implement support for getting the platform, you will *always* have to run at least 2 commands.

So basically what I am saying is that in general the fallback mechanism is more efficient than targeting based on platform.

Comment (dagwieers) [reply lib/ansible/plugins/action/reboot.py:112]: And yes, PowerShell has aliases for common Unix commands, which is freaking a lot of people out (because they are not drop-in replacements). https://daniel.haxx.se/blog/2016/08/19/removing-the-powershell-curl-alias/

**Review by samdoran — APPROVED** [review]

File: lib/ansible/plugins/action/reboot.py [hunk 4]
```diff
@@ -94,18 +94,31 @@ def construct_command(self):
         return reboot_command
 
     def get_system_boot_time(self):
+        stdout = ''
+        stderr = ''
```
Comment (bcoca) [inline lib/ansible/plugins/action/reboot.py:98]: this might not be py2/py3 compatible, we might want to force bytes

Comment (samdoran) [reply lib/ansible/plugins/action/reboot.py:98]: Yes, you are correct. One day I'll master that distinction. 😄

**Comment by dagwieers** [comment]
@samdoran I assumed your own PR would take over this. That's why I didn't bother doing the fragment.

**Comment by samdoran** [comment]
@dagwieers I wanted to get this fix in. My PR is probably too big a change to get in the short term, plus it's probably in the feature request territory.

**Comment by dagwieers** [comment]
Ok. Sorry for the confusion.

**Comment by samdoran** [comment]
Not at all. Thank you for finding and fixing this.
